src/relayer/publicDecrypt.ts

```diff
@@ -1,28 +1,170 @@
 import { fromHexString, toHexString } from '../utils';
-import { ethers } from 'ethers';
+import { ethers, AbiCoder } from 'ethers';
+
+export type DecryptedResults = Record<string, bigint | boolean | string>;
 
 const aclABI = [
-  'function persistAllowed(bytes32 handle, address account) view returns (bool)',
+  'function isAllowedForDecryption(bytes32 handle) view returns (bool)',
 ];
 
+function isThresholdReached(
+  kmsSigners: string[],
+  recoveredAddresses: string[],
+  threshold: number,
+): boolean {
+  const addressMap = new Map<string, number>();
+  recoveredAddresses.forEach((address, index) => {
+    if (addressMap.has(address)) {
+      const duplicateValue = address;
+      throw new Error(
+        `Duplicate signer address found: ${duplicateValue} appears multiple times in recovered addresses`,
+      );
+    }
+    addressMap.set(address, index);
+  });
+
+  for (const address of recoveredAddresses) {
+    if (!kmsSigners.includes(address)) {
+      throw new Error(
+        `Invalid address found: ${address} is not in the list of KMS signers`,
+      );
+    }
+  }
+
+  return recoveredAddresses.length >= threshold;
+}
+
+const NumEncryptedBits: Record<number, number> = {
+  0: 2, // ebool
+  2: 8, // euint8
+  3: 16, // euint16
+  4: 32, // euint32
+  5: 64, // euint64
+  6: 128, // euint128
+  7: 160, // eaddress
+  8: 256, // euint256
+  9: 512, // ebytes64
+  10: 1024, // ebytes128
+  11: 2048, // ebytes256
+} as const;
+
+function checkEncryptedBits(handles: string[]) {
+  let total = 0;
+
+  for (const handle of handles) {
+    if (handle.length !== 66) {
+      throw new Error(`Handle ${handle} is not of valid length`);
+    }
+
+    const hexPair = handle.slice(-4, -2).toLowerCase();
+    const typeDiscriminant = parseInt(hexPair, 16);
+
+    if (!(typeDiscriminant in NumEncryptedBits)) {
+      throw new Error(`Handle ${handle} is not of valid type`);
+    }
+
+    total +=
+      NumEncryptedBits[typeDiscriminant as keyof typeof NumEncryptedBits];
+
+    // enforce 2048‑bit limit
+    if (total > 2048) {
+      throw new Error(
+        'Cannot decrypt more than 2048 encrypted bits in a single request',
+      );
+    }
+  }
+  return total;
+}
+
+const CiphertextType: Record<number, 'bool' | 'uint256' | 'address' | 'bytes'> =
+  {
+    0: 'bool',
+    2: 'uint256',
+    3: 'uint256',
+    4: 'uint256',
+    5: 'uint256',
+    6: 'uint256',
+    7: 'address',
+    8: 'uint256',
+    9: 'bytes',
+    10: 'bytes',
+    11: 'bytes',
+  };
+
+function deserializeDecryptedResult(
+  handles: string[],
+  decryptedResult: string,
+): Record<string, DecryptedResults> {
```

Comment:
You should return `DecryptedResults` instead of `Record<string, DecryptedResults>`